### PR TITLE
login working

### LINKED ISSUE #8: Backend: add authentication
Maybe steal from:
https://github.com/transitiverobotics/transitive/blob/main/cloud/app/server/server.js#L47-L79

### PR DESCRIPTION
**Pull request by jlurgo**
Login and logout are working!

Pending:
- stay logged in when refreshing the page, not working currently
- make login view nicer
- user management for admins?
- recover and change password
- register

image
image

closes #8

### CODE REVIEW

File: sample.env [hunk 1]
```diff
@@ -9,4 +9,16 @@ VITE_TRANSITIVE_USER=
 
 # The JWT secret of the above user on the above Transitive deployment's
 # Security page
-JWT_SECRET=
\ No newline at end of file
+JWT_SECRET=
+
+# Secret used by express-session
+TR_SESSION_SECRET=SOME_STRONG_SECRET_LIKE_CH9UwdjGck4DKGMbftVbBm4r9SuZQKDP
```
Comment (chfritz) [inline sample.env]: Let's not use the `TR_` prefix here. That's short for Transitive Robotics and we are using it for Transitive itself, so let's not overload it. Maybe we can use `TRANSACT_`?

File: sample.env [hunk 2]
```diff
@@ -9,4 +9,16 @@ VITE_TRANSITIVE_USER=
...
+
+# Used at build time. For local development we recommend YOUR_HOSTNAME.local
+TR_HOST=your-domain.com
```
Comment (chfritz) [inline sample.env]: I know we are using it in the Transitive portal server, but I think here we may not need this.

File: src/server/accounts.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "_id": "admin",
+    "bcryptPassword": "$2b$10$o78d5eRDsumu7L.Uq7WlQO2imW/9k7VWateMa/4YOu7Yxad4yF15m",
```
Comment (chfritz) [inline src/server/accounts.json]: We probably shouldn't include a default password. Can we remove this file?

File: sample.env [hunk 4]
```diff
@@ -9,4 +9,16 @@ VITE_TRANSITIVE_USER=
...
+
+# create this account in accounts.json file if it doesn't exist
+TR_USER=superadmin
+TR_PASS=some-very-secure-password
+TR_EMAIL=admin_@superrobots.com
```
Comment (chfritz) [inline sample.env]: Can we change our branding to "SuperBots"? It rolls of the tongue more easily than superrobots.

File: src/server/main.ts [hunk 5]
```diff
@@ -2,26 +2,110 @@ import express from "express";
...
 
 const app = express();
 app.use(express.json());
 
+// Set up session middleware
+app.use(session({
+  secret: process.env.TR_SESSION_SECRET, // used to sign the session ID cookie
+  resave: false,
+  saveUninitialized: true,
+  cookie: {domain: `.${process.env.TR_HOST}`, sameSite: 'strict'}
```
Comment (chfritz) [inline src/server/main.ts]: We may not need this here, and can then avoid the need for TR_HOST.

**Review by chfritz — COMMENTED** [review]
Looks good so far. Just some comments. But let's only merge this once the page is working again (right now the sidebar works, but the body doesn't).

File: sample.env [hunk 6]
```diff
@@ -9,4 +9,16 @@ VITE_TRANSITIVE_USER=
...
+JWT_SECRET=
+
+# Secret used by express-session
+TR_SESSION_SECRET=SOME_STRONG_SECRET_LIKE_CH9UwdjGck4DKGMbftVbBm4r9SuZQKDP
+
+# Used at build time. For local development we recommend YOUR_HOSTNAME.local
+TR_HOST=your-domain.com
+
+# create this account in accounts.json file if it doesn't exist
+TR_USER=superadmin
```
Comment (chfritz) [inline sample.env]: I think "admin" is sufficient.

File: sample.env (hunk 4 above)
Comment (jlurgo) [reply sample.env]: Yes! That double RR made me itchy

File: src/server/accounts.json (hunk 3 above)
Comment (jlurgo) [reply src/server/accounts.json]: That's the admin user created on server startup. I can add this file to git ignore and make it to be created if not exists

**Review by chfritz — APPROVED** [review]

**Pull request merged**